src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -1228,6 +1229,121 @@ func (s *service) WriteBatchRaw(tctx thrift.Context, req *rpc.WriteBatchRawReque
 	return nil
 }
 
+func (s *service) WriteBatchRawV2(tctx thrift.Context, req *rpc.WriteBatchRawV2Request) error {
+	s.metrics.writeBatchRawRPCs.Inc(1)
+	db, err := s.startWriteRPCWithDB()
+	if err != nil {
+		return err
+	}
+	defer s.writeRPCCompleted()
+
+	callStart := s.nowFn()
+	ctx := tchannelthrift.Context(tctx)
+
+	// Sanity check input.
+	numNamespaces := int64(len(req.NameSpaces))
+	for _, elem := range req.Elements {
+		if elem.NameSpace >= numNamespaces {
+			return fmt.Errorf("namespace index: %d is out of range of provided namespaces", elem.NameSpace)
+		}
+	}
+
+	// Sort the elements so that they're sorted by namespace so we can reuse the same batch writer.
+	sort.Slice(req.Elements, func(i, j int) bool {
+		return req.Elements[i].NameSpace < req.Elements[j].NameSpace
+	})
+
+	// NB(r): Use the pooled request tracking to return thrift alloc'd bytes
+	// to the thrift bytes pool and to return ident.ID wrappers to a pool for
+	// reuse. We also reduce contention on pools by getting one per batch request
+	// rather than one per ID.
+	pooledReq := s.pools.writeBatchPooledReqPool.Get()
+	pooledReq.writeV2Req = req
+	ctx.RegisterFinalizer(pooledReq)
+
+	var (
+		nsID        ident.ID
+		nsIdx       int64
+		batchWriter ts.BatchWriter
+
+		retryableErrors    int
+		nonRetryableErrors int
+	)
+	for i, elem := range req.Elements {
+		if nsID == nil || elem.NameSpace != nsIdx {
+			if batchWriter != nil {
+				err = db.WriteBatch(ctx, nsID, batchWriter.(ts.WriteBatch), pooledReq)
+				if err != nil {
+					return convert.ToRPCError(err)
+				}
+				batchWriter = nil
+			}
+
+			nsID = s.newPooledID(ctx, req.NameSpaces[elem.NameSpace], pooledReq)
+			nsIdx = elem.NameSpace
+
+			batchWriter, err = db.BatchWriter(nsID, len(req.Elements))
+			if err != nil {
+				return convert.ToRPCError(err)
+			}
+			// The lifecycle of the annotations is more involved than the rest of the data
+			// so we set the annotation pool put method as the finalization function and
+			// let the database take care of returning them to the pool.
+			batchWriter.SetFinalizeAnnotationFn(finalizeAnnotationFn)
+		}
+
+		unit, unitErr := convert.ToUnit(elem.Datapoint.TimestampTimeType)
+		if unitErr != nil {
+			nonRetryableErrors++
+			pooledReq.addError(tterrors.NewBadRequestWriteBatchRawError(i, unitErr))
+			continue
+		}
+
+		d, err := unit.Value()
+		if err != nil {
+			nonRetryableErrors++
+			pooledReq.addError(tterrors.NewBadRequestWriteBatchRawError(i, err))
+			continue
+		}
+
+		seriesID := s.newPooledID(ctx, elem.ID, pooledReq)
+		batchWriter.Add(
```

Comment:
Can `batchWriter` be nil at this stage?

Reply:
No I don't think so. It starts off as nil and it will get set in the first iteration of the loop. Everytime after that where it gets set to nil (caus a batch was written) we assign a new one